Add caching of AclFileAttributeView as a property to AttributeCachingPath

mbayerPK opened the pull request:
Add usage of the AttributeCachingPath cached AclFileAttributeView property to AttributeCachingFileSystemProvider functions

Add tests in AttributeCachingFileSystemTests for acl attributes and AclFileAttributeView

Modify some file attribute tests in AttributeCachingFileSystemTests to use the default filesystem as JimFs does not properly emulate posix, dos, and acl file attributes

PKJosh commented:
@mbayerPK  Just letting you know I haven't looked at this because there are failing tests. Didn't want to ping you if you already knew and were already trying to fix it.

PKJosh commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingFileSystemProvider.kt:
```diff
@@ -224,7 +225,14 @@ internal class AttributeCachingFileSystemProvider : FileSystemProvider() {
     ): V {
         val actualPath = path.getUnderlyingPath()
-        return actualPath.fileSystem.provider().getFileAttributeView(actualPath, type, *options)
+        return if (path is AttributeCachingPath && type is AclFileAttributeView) {
```
Why are we handling this any differently than we would other views? `actualPath.fileSystem.provider().getFileAttributeView(actualPath, type, *options)` should work. And if we want the benefits of reading from cache, then the other attributes should probably also be doing the same thing.

mbayerPK replied:
so for acl file attributes, I am caching the entire AclFileAttributeView rather than the attributes- this is very different than how the other attributes are currently cached (as they all extend BasicFileAttributes - which is not the view)

I have to cache the AclFileAttributeView because its difficult to translate the UserPrincipal owner and the AclEntry list by themselves to cachable variables (they'd have to be separate cached values). 
You cannot access the acl attributes via the provider function `fun <A : BasicFileAttributes?> readAttributes(path: Path,
        type: Class<A>,
        vararg options: LinkOption)` because thats for attributes that extend BasicFileAttributes only.
You can access acl attributes via `fun <V : FileAttributeView?> getFileAttributeView(}` or by `fun readAttributes(
        path: Path,
        attributes: String,
        vararg options: LinkOption
    )`

I need to do other_iops testing to see if caching the view truly works for performance, or if there is a performance hit everytime you query the view because the underlying filesystem is accessed to get the actual attributes.

PKJosh commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingFileSystemProvider.kt:
```diff
@@ -370,14 +378,21 @@ internal class AttributeCachingFileSystemProvider : FileSystemProvider() {
         // attribute to properly set the cache.
-        val attributesObject = getAttributesClassFromPathProvider(path, attributeClassName)
-        path.setAttributeByName(attributeClassName, attributesObject)
+        val attributesView: FileAttributeView? = getAttributesClassFromPathProvider(path, attributeClassName)
+        if (attributeClassName == "acl:*") {
+            path.aclFileAttributeView = attributesView as? AclFileAttributeView
+        } else {
+            val basicFileAttributesView = attributesView as? BasicFileAttributeView
+            path.setAttributeByName(attributeClassName, basicFileAttributesView?.readAttributes())
+        }
```
`AclFileAttributeView` is a `FileAttributeView`, so I think handling acl in `setAttributeByName` is perfectly appropriate and simplifies the logic.

mbayerPK replied:
SetAttributeByName takes in a BasicFileAttributes object, not the view, so we cannot do this

mbayerPK replied:
okay i see what you mean this makes sense and id just edit what happens within setAttributeViewByName

PKJosh commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingPath.kt:
```diff
@@ -47,13 +49,28 @@ internal class AttributeCachingPath(
+     * Once the [AclFileAttributeView] is cached it can only be modified/re-cached via [Files.setAttribute] or other
+     * attribute setting methods in [[FileSystemProvider].
+     */
+    var aclFileAttributeView by LazyAttribute {
```
If my other suggestion is followed, this could just be `private var`.

mbayerPK replied:
See above replies

PKJosh commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingPath.kt:
```diff
@@ -210,6 +226,11 @@ internal class AttributeCachingPath(
             target.basicAttributes = basicFileAttributes
             target.dosAttributes = dosFileAttributes
             target.posixAttributes = posixFileAttributes
+            target.aclFileAttributeView = if (supportedViews.contains("acl")) {
```
This could just be `target.aclFileAttributeView = aclFileAttributeView`, as it's nullable like the rest.

mbayerPK replied:
this follows the same check logic above on lines 217-223 for the dos and posix file attributes to check if the view is supported

mbayerPK replied:
yeah i went with your shortened version ryan and added the supportedFileViews lazy attribute, I think I did it the original way to save other_iops hits with calling delegate.fileSystem.supportedFileAttributeViews() once instead of for each lazy attribute that was null.

PKJosh commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingPath.kt:
```diff
@@ -256,44 +277,53 @@ internal class AttributeCachingPath(
-            dosAttributes
-        } else if (checkedNames.startsWith("posix")) {
-            posixAttributes
+        if (checkedNames.startsWith("acl")) {
```
This could just be part of the same if/else ladder that existed before

mbayerPK replied:
it cannot because we are working with the acl file attribute view vs an object that extends BasicFileAttributes

mbayerPK commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingFileSystemProvider.kt:
```diff
@@ -224,7 +225,14 @@ internal class AttributeCachingFileSystemProvider : FileSystemProvider() {
+        return if (path is AttributeCachingPath && type is AclFileAttributeView) {
+            @Suppress("UNCHECKED_CAST")
+            path.aclFileAttributeView as V ?: throw UnsupportedOperationException(
+                "Could not read file attribute view $type from the path $path and its delegate filesystem."
```
this i could do pretty easily

mbayerPK commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingFileSystemProvider.kt:
```diff
@@ -218,13 +220,31 @@ internal class AttributeCachingFileSystemProvider : FileSystemProvider() {
+            @Suppress("UNCHECKED_CAST")
+            path.aclFileAttributeView as V
+        } else {
+            actualPath.fileSystem.provider().getFileAttributeView(actualPath, type, *options)
```
yeah umm then this becomes very very complicated as these objects are completely different than the other file attributes.

PKJosh replied:
Just as an update, while I attached a screenshot of a JIMFS class, that as I suspected it might, the Windows implementation unfortunately shares this problem.

<img width="761" alt="image" src="https://github.com/pkware/AttributeCachingFileSystem/assets/77342611/5caf8ed2-0e7e-45c9-b42e-4420f766f377">

mbayerPK replied:
yeah i am just going with caching the individual acl attributes, not going to make a separate class to hold them both though- theres no need

PKJosh replied:
> yeah i am just going with caching the individual acl attributes, not going to make a separate class to hold them both though- theres no need

Agreed, I was just using that point to illustrate that it is trivial to accomplish. Absolutely fine to do it the way you suggested.

mbayerPK commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingFileSystemProvider.kt:
```diff
@@ -218,13 +220,31 @@ internal class AttributeCachingFileSystemProvider : FileSystemProvider() {
     ): V {
         val actualPath = path.getUnderlyingPath()
-        return actualPath.fileSystem.provider().getFileAttributeView(actualPath, type, *options)
+        return if (path is AttributeCachingPath && type is AclFileAttributeView) {
```
i ended up getting rid of this logic when caching the individual acl attributes

mbayerPK commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingFileSystemProvider.kt:
```diff
@@ -218,13 +220,31 @@ internal class AttributeCachingFileSystemProvider : FileSystemProvider() {
         type: Class<V>?,
         vararg options: LinkOption?
     ): V {
         val actualPath = path.getUnderlyingPath()
```
i ended up getting rid of this logic when caching the individual acl attributes

mbayerPK commented on file-attribute-caching/src/test/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingFileSystemTests.kt:
```diff
@@ -350,27 +358,37 @@ class AttributeCachingFileSystemTests {
+    @DisabledOnOs(OS.WINDOWS)
+    fun `set and read posix attributes for path`() = AttributeCachingFileSystem.wrapping(FileSystems.getDefault()).use {
+        // get filesystem attribute caching path
+        val javaTmpDir = it.getPath(System.getProperty("java.io.tmpdir"))
```
so the issue here was that / calls resolve with a string and in those cases we were not returning caching paths, i added tests for those and cleared out these asserts

there was a second issue where the final caching path never gets initialized after the file is made so I forced it to become initialized and added asserts around that because the acl/dos tests depend on the cache being setup properly

PKJosh commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingFileSystemProvider.kt:
```diff
@@ -44,7 +44,8 @@ internal class AttributeCachingFileSystemProvider : FileSystemProvider() {
     private val isWindowsOS by lazy {
         // Adapted from org/junit/jupiter/api/condition/OS.java to look up the operating system name
-        System.getProperty("os.name").lowercase(Locale.ENGLISH).contains("win")
+        // lowercase() defaults to Locale.ROOT in kotlin, but toLowerCase() in java does not, BE AWARE OF THIS.
```
Still wondering who this comment is _for_.

We're not migrating this code to Java, readers of this library can suss this information for themselves, and we've already handled the safe operation of this for them. What is there for us to warn them about?

mbayerPK replied:
we can agree to disagree here, I am leaving the warning in as you literally have to dig into the code to see whats happening behind the scenes in kotlin vs java. I shouldn't have to do that to understand whats happening at a glance.

PKJosh commented on file-attribute-caching/src/main/kotlin/com/pkware/filesystem/attributecaching/AttributeCachingPath.kt:
```diff
@@ -162,22 +219,37 @@ internal class AttributeCachingPath(
         // This check is to ensure that we are only storing attribute classes and not specific attributes.
         // Remove basic: from our attribute name if present as basicFileAttributes can be accessed without that
         // qualifier
         when (name.substringAfter("basic:")) {
```
I still wish Java had something like an attribute parsing component for this or ours was a little smarter.

We technically allow input like `basic:posix:group` or `basic:acl:owner`, returning the properties without error. It's harmless but strange.

mbayerPK committed:
Add caching of AclFileAttributeView owner and aclEntries as properties to AttributeCachingPath

Add function overrides to  AttributeCachingPath for resolve and resolveSibling when called with strings

Modify AttributeCachingFileSystemProvider functions to support AclFileAttributeView

Add tests in AttributeCachingFileSystemTests for acl attributes

Add tests in AttributeCachingFileSystemTests for resolve and resolveSibling when called with strings

Modify some file attribute tests in AttributeCachingFileSystemTests to use the default filesystem as JimFs does not properly emulate posix, dos, and acl file attributes
